### PR TITLE
feat: resolve $refs in preprocessors

### PR DESCRIPTION
**Pull request by tatomyr**
## What/Why/How?

> TODO: update docs

Resolve $refs after running preprocessors with the following option in **redocly.yaml**:

```yaml
...
resolve:
   resolveAfterTransformers: true
```

## Reference

Closes https://github.com/Redocly/redocly-cli/issues/324

## Check yourself

- [x] Code is linted
- [x] Tested with redoc/reference-docs/workflows
- [x] All new/updated code is covered with tests

## Security

- [x] Security impact of change has been considered
- [x] Code follows company security practices and guidelines

### CODE REVIEW

File: __tests__/bundle/bundle-lint-format/checkstyle-format-snapshot.js [hunk 1]
```diff
@@ -9,7 +9,7 @@ exports[`E2E bundle lint format bundle lint: should be formatted by format: chec
 </checkstyle>
 
 ❌ Validation failed with 1 error.
-run \`openapi lint --generate-ignore-file\` to add all problems to the ignore file.
+run \`redocly lint --generate-ignore-file\` to add all problems to the ignore file.
```
Comment (RomanHotsiy) [inline __tests__/bundle/bundle-lint-format/checkstyle-format-snapshot.js:12]: Could you extract those changes into separate PR plz?

Comment (tatomyr) [reply __tests__/bundle/bundle-lint-format/checkstyle-format-snapshot.js:12]: Sure. Just stored everything in the draft in order not to lose the changes.

**Comment by rotorgames** [comment]
@tatomyr will we be able to set it up in the redocly.config to make it work with the vscode extension?

File: packages/core/src/bundle.ts [hunk 2]
```diff
@@ -129,6 +129,23 @@ export async function bundleDocument(opts: {
     });
   }
 
+  if (config.resolveAfterTransformers) {
```
Comment (RomanHotsiy) [inline packages/core/src/bundle.ts:132]: I was thinking about it a bit. Let's not have this config.

Let's enable it if only there are any preprocessors enabled.

We don't have too many users using preprocessors I believe.
We should just add a little warning somewhere in the docs that using preprocessors has a performance penalty.

File: packages/core/src/lint.ts [hunk 3]
```diff
@@ -77,13 +77,35 @@ export async function lintDocument(opts: {
 
   const preprocessors = initRules(rules as any, config, 'preprocessors', oasVersion);
   const regularRules = initRules(rules as Oas3RuleSet[], config, 'rules', oasVersion);
-  const normalizedVisitors = normalizeVisitors([...preprocessors, ...regularRules] as any, types);
+
+  if (config.resolveAfterTransformers) {
```
Comment (RomanHotsiy) [inline packages/core/src/lint.ts:81]: I think this is good to just have this piece copied over. Extracting into function may be tempting but this is easier to read I think.

👍

**Review by RomanHotsiy — COMMENTED** [review]
Looks pretty good.

Left a few comments.

File: packages/core/src/bundle.ts [hunk 4]
```diff
@@ -129,6 +129,23 @@ export async function bundleDocument(opts: {
...
+    // Make preliminary pass to be able to resolve refs defined in preprocessors in the next pass.
+    const preliminaryResolvedRefMap = await resolveDocument({
```
[the remark this answers is not in the document]
Comment (tatomyr) [reply packages/core/src/bundle.ts:134]: Could be. Let me rewrite it without the config option and see if that would be better.

**Comment by tatomyr** [comment]
@rotorgames yes, I believe it will work in VS Code extension by default.

File: packages/core/src/bundle.ts (hunk 4 above)
[the remark this answers is not in the document]
Comment (tatomyr) [reply packages/core/src/bundle.ts]: Yeah, using `let` makes more sense in this context. Refactored.

**Review by IgorKarpiuk — APPROVED** [review]
